### PR TITLE
Add Javascript and TypeScript build for jetpack-wpcom-mu

### PR DESCRIPTION
**Pull request by daledupreez**
## Proposed changes:

* This PR adds initial Javascript and TypeScript build logic for the jetpack-mu-wpcom package.
* The build logic relies on `webpack` and Jetpack's default Webpack configuration
* The PR folds in the work from #34159 to build the code from `src/features/error-reporting`, a project we want to migrate from the Editing Toolkit Plugin in https://github.com/Automattic/wp-calypso/tree/05bbf6a425badcdc73cc9ff27c0fc418c018fb2b/apps/editing-toolkit

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

No product changes for now -- we simply want to prepare for pulling more features from the WordPress.com Editing Toolkit plugin so they're available in a single location across WordPress.com Simple and Atomic.

## Does this pull request change what data or activity we track or use?

No changes to data or activity tracking.

### Local testing

* Check out this branch locally
* Run `pnpm i`
* Run `pnpm run build-js`
* Verify that you get JS output in `src/build/error-reporting/error-reporting.js`, and that you get `src/build/error-reporting/error-reporting.asset.php` and `src/build/error-reporting/error-reporting.js.map` produced
* Run `pnpm run build-production-js`
* Verify that `src/build/error-reporting/` now only contains `error-reporting.js` (which should be minified) and `error-reporting.asset.php`

### Functional testing - WordPress.com Simple

* On a WordPress.com sandbox, run `bin/jetpack-downloader test jetpack-mu-wpcom-plugin add/js-build-for-jetpack-wpcom-mu` (as documented in [this comment](https://github.com/Automattic/jetpack/pull/34158#issuecomment-1814090289)) to apply this code to your sandbox
* Ensure that requests for both `s0.wp.com` and your development site are routed via your WordPress.com sandbox
* Add the following lines to your `0-sandbox.php` file to force file concatenation and :
```php
define( 'FORCE_CONCAT', TRUE );
add_filter( 'wpcom_sandbox_staticize_subdomain', '__return_true' );
```
* From a command line on your sandbox, run the following command to add the `error-reporting-use-sentry` sticker to your site: `wp blog-stickers add --blog_id=[YOUR_BLOG_ID] --sticker=error-reporting-use-sentry --who=[YOUR_USERNAME]`
* Navigate to `/wp-admin/edit.php` for your test site
* Click on "Edit" for one of your posts (or create a new one if you need to)
* Verify that you are viewing the post editor via `/wp-admin/post.php` (i.e. we want to make sure you're _not_ viewing the post editor via `wordpress.com/post/`)
* Open up the developer console, and enter the following code in the console:
```js
var testETK = document.createElement('script');
testETK.innerHTML = "throw new Error('ETK - hi there!');";
document.body.appendChild( testETK );
```
* Check your network tab and verify that we make an HTTP request to a subdomain of `sentry.io`
* Now disable the ETK code on your sandbox by commenting out the following function call in `wp-content/plugins/editing-toolkit-plugin/prod/error-reporting/index.php`:
```php
if ( is_site_eligible_for_error_reporting() ) {
	//setup_error_reporting();
}
```
* Reload your editor page
* Verify that the HTML source for the editor page includes a script with the following source: `https://s0.wp.com/wp-content/mu-plugins/jetpack-mu-wpcom-plugin/moon/vendor/automattic/jetpack-mu-wpcom/src/build/error-reporting/error-reporting.js`
* Open up the developer console, and enter the following code in the console:
```js
var testJetpackMuWPCOM = document.createElement('script');
testJetpackMuWPCOM.innerHTML = "throw new Error('JPMUWPCOM - hi there!');";
document.body.appendChild( testJetpackMuWPCOM );
```
* Check your network tab and verify that we make an HTTP request to a subdomain of `sentry.io`

### Functional testing - WPCOM Atomic

* Ensure you have the Jetpack Beta plugin installed on your Atomic site. (You can download the current version from https://jetpack.com/download-jetpack-beta/.)
* Navigate to _Jetpack_ -> _Beta Tester_ in the left nav on your site
* Click on the "Manage" button for the _WordPress.com Features_ item
* Type `add/js-build-for-jetpack-wpcom-mu` in the search field, and then click "Activate" for the version from this branch
* Follow the instructions in [this comment](https://github.com/Automattic/jetpack/pull/34158#issuecomment-1814090289) to run the code from the beta plugin
* Navigate to `/wp-admin/edit.php` for the test site
* Click on "Edit" for one of your posts (or create a new one if you need to)
* * Open up the developer console, and enter the following code in the console:
```js
var testETK = document.createElement('script');
testETK.innerHTML = "throw new Error('ETK - hi there!');";
document.body.appendChild( testETK );
```
* Check your network tab and verify that we did _not_ make an HTTP request to a subdomain of `sentry.io` - we don't have this feature enabled for Atomic sites at this time

### CODE REVIEW

**Comment by obenland** [comment]
I did a functional test following the instructions in #54257 and it looks like it works as expected. The only thing I noticed missing was the assets file that contains dependencies etc

**Comment by okmttdhr** [comment]
I'm just done with `Local testing`, and that worked well :)

File: projects/packages/jetpack-mu-wpcom/src/features/error-reporting/error-reporting.php
Comment (okmttdhr) [inline projects/packages/jetpack-mu-wpcom/src/features/error-reporting/error-reporting.php:97]: Am I correct in understanding that this feature is disabled because user_id on Atomic is not linked to WordPress.com? I'm asking for my learning purposes 🙂 @daledupreez @obenland 
PCYsg-Ewq-p2

**Review by okmttdhr — DISMISSED** [review]
I did `Functional testing` on Simple and Atomic site, and it worked well! 🎉 Thank you so much for your work. 

📝 
- I also merged `trunk` into this branch. 
- The assets was served by `<script crossorigin="anonymous" type="text/javascript" src="https://s0.wp.com/_static/??-eJx9T0kOwjAM/BDGLUIgDoi3uIlVUppFjkN5PhFFCIrgZs/iGeOUwMSgHBTJehcgjaV3IaNn6wj4lsYoLJCVNK+HvMI3hy8v+WzuSNBTVpY6gQqZyz/TwJqqBCo0JRP9k8BcAl452ChIRaMnVWe+1JjFYFfcaJFFooBwiqIu9Mt9WeHz0zn6URZm5sxj+gHXUyd/bPdN024Pu81muANc1ni8"></script>` in my case.

**Review by okmttdhr — DISMISSED** [review]

File: projects/packages/jetpack-mu-wpcom/src/features/error-reporting/error-reporting.php
Comment (daledupreez) [reply projects/packages/jetpack-mu-wpcom/src/features/error-reporting/error-reporting.php:97]: @okmttdhr, I think that @fullofcaffeine is best-placed to comment on why we're only running this on WPCOM Simple and haven't rolled the functionality out more broadly.

**Comment by alshakero** [comment]
Hi Dale! We're really excited about moving Verbum into this. Do you have an idea when it will be deployed?

**Comment by daledupreez** [comment]
@alshakero, we're currently blocked on a functional review of the Sentry code by @fullofcaffeine.

@fullofcaffeine, can you try to prioritise the review so we can start work on our next migrations that depend on the build code in this PR?

**Comment by fullofcaffeine** [comment]
@daledupreez Hi! I'm checking this out now; sorry for the delay, had a busy week.

File: projects/packages/jetpack-mu-wpcom/src/features/error-reporting/error-reporting.php
Comment (fullofcaffeine) [reply projects/packages/jetpack-mu-wpcom/src/features/error-reporting/error-reporting.php:97]: Hi @okmttdhr! Yes, it's not running on AT. The original plan is to run on AT at some point, and we tried - see p4TIVU-9DI-p2#comment-10922 - but there were some problems. At the time, though, we hadn't implemented Sentry yet, and Sentry should work fine on AT, too. The challenge, though is how to treat errors from AT as AT is the wild west :D - sites can run all sorts of third-party code - so we should have a RFC before proceeding, and Sentry atm is not top-priority for WPCOM, AFAIK.

**Review by fullofcaffeine — DISMISSED** [review]
LGMT, and tests well! I also tested logstash capture (when Sentry is disabled) and it works great. The negative scenario for AT also tests well.

Now that this will live in the `jetpack-wpcom-mu`, where can I find more info about deploying it on my own? Or will I need to coordinate JP release with the JP folks?

The only missing piece will be to remove the `error-reporting-module` from ETK - let me know if you need a hand there. I think we could ship it just after shipping this one.

Thanks!

**Comment by daledupreez** [comment]
Thanks for testing and taking a look, @fullofcaffeine!

> Now that this will live in the `jetpack-wpcom-mu`, where can I find more info about deploying it on my own? Or will I need to coordinate JP release with the JP folks?

At present, `jetpack-mu-wpcom` is deployed to WordPress.com Simple once a day, and to WordPress.com Atomic once a week. It's also possible to release separately to both platforms. We have a lot more documentation about the release process and development worklow here: PCYsg-Osp-p2

> The only missing piece will be to remove the error-reporting-module from ETK - let me know if you need a hand there. I think we could ship it just after shipping this one.

That's my plan -- I'll aim to get this PR shipped on Monday and get the removal PR going then as well.

**Review by miksansegundo — APPROVED** [review]
🚀

**Pull request merged**
